### ISSUE TITLE
R3: Plan-driven transform dispatch Implementation Plan

### ISSUE THREAD

**Issue by JueonPark** [JueonPark/sym#147]
**Goal:** Execute explicitly requested typed relocation programs through qualified CPU/CUDA paths, with observable dispatch and a forceable original CPU-transform pipeline.

**Architecture:** Consume C3's typed bound plan and R2's validated transport rather than inferring computation from a cost-model decision. Enumerate only implemented, semantically equivalent execution paths; reuse existing quant/cast kernels when their contracts match and supply scalar reference execution for the initial typed subset. Cost estimates rank available paths without granting capability.

**Tech Stack:** Existing standalone C++ runtime, CPU transforms, CUDA 12.6 kernels and backend; optional pybind11 3.0.4 / CPython 3.14.7 / PyTorch 2.14.0 bridge. Preserve the compiler's LLVM/MLIR 21.1.8 pin and the existing runtime language settings.

**Spec:** [#133](https://github.com/JueonPark/sym/issues/133), [project plan §3/R3](https://github.com/JueonPark/sym/blob/main/docs/project-finalization-plan.md#3-complete-compiler-to-runtime-execution), [C1 numerical semantics](https://github.com/JueonPark/sym/issues/141), [C2 ordered folding](https://github.com/JueonPark/sym/issues/142), [C3 typed bindings](https://github.com/JueonPark/sym/issues/143), and [single cost-model implementation](https://github.com/JueonPark/sym/issues/111).

## Global Constraints

- "Keep torch dependencies outside the core runtime."
- Cast, quantize, and dequantize are explicit requested operations; runtime placement cannot silently introduce lower precision or erase lossy stages.
- Separate source, selected wire, destination, and parameter byte footprints. Derive transfer storage from an actual legal stage partition, never a guessed `wireRatio`.
- Missing calibration changes placement advice, not numerical semantics or capability.
- D2H executes a forward value/layout program. Quantization and narrowing casts have no assumed inverse.
- Preserve v0 layout execution. Typed v1 binding alone does not qualify a kernel; existing `BoundPlan.decision` remains advice until consumed by this dispatcher.
- Initial nonblocking behavior follows R2's qualified completion contract. Int4 and training/calibration algorithms are outside this typed scope.

## Review Focus

- An existing kernel can disagree with C1 on rounding, NaNs, zero points, or channel indexing (Tasks 1–2).
- Pad-before-quantize and quantize-before-pad can require different values and parameter access (Tasks 1–2).
- A cost estimate may recommend a path whose kernel or exact stage partition is unavailable (Task 3).
- Invalid or changed runtime parameter values must fail or invalidate preparation before data execution (Tasks 1 and 4).
- Prefolded data can retain valid addresses while its weights or parameters have changed (Task 4).

---

**Proposed child title:** `[Finalize][Runtime][R3] Dispatch typed plans through qualified transforms`

**Dependencies:** [C1](https://github.com/JueonPark/sym/issues/141), [C2](https://github.com/JueonPark/sym/issues/142), and [C3](https://github.com/JueonPark/sym/issues/143) provide the numerical, ordered-plan, and versioned binding contracts; [R2](https://github.com/JueonPark/sym/issues/146) provides validated buffers, forward transport, and lifetime/stream handling. [C4 / #144](https://github.com/JueonPark/sym/issues/144) consumes [R3](https://github.com/JueonPark/sym/issues/147) execution for conformance, so [R3](https://github.com/JueonPark/sym/issues/147) must not depend on C4 completion. Share C1's numerical vectors and coordinate C4's independent oracle early. [T4 / #137](https://github.com/JueonPark/sym/issues/137) owns the weight/prefold frontend integration.

**Runtime subissues:** [R1](https://github.com/JueonPark/sym/issues/145), [R2](https://github.com/JueonPark/sym/issues/146), [R3](https://github.com/JueonPark/sym/issues/147), [R4](https://github.com/JueonPark/sym/issues/148).

## Starting evidence

At main `4e67545d2e6331802a7e295c65c135cf90b26a4f`, CPU/CUDA quant, cast, dequant, and prefold primitives exist, but there is no unified plan-driven dispatcher. Existing S8 quantization uses inverse scales, nearest-even rounding and saturation, no zero point, and a specific NaN result. Gather/quant and dequant/relocate kernels have layout/channel/rank restrictions. The runtime has a single element-size layout contract; C3 must land distinct typed descriptors/footprints before these kernels can safely consume typed plans.

`PrefoldArtifact` currently retains a raw backend pointer, and `OutputSpec` contains only `S8GatherQuant`/`S8QuantPack`. Neither a generic f32 prefold path nor a backend lifetime guarantee can be inferred from this API.

## File map

| Action | Path | Responsibility |
| --- | --- | --- |
| Create | `libreloc/include/reloc/Dispatch.h`, `libreloc/src/Dispatch.cpp` | Capability admission, explicit policy, selected stage partition, execution report |
| Create | `libreloc/include/reloc/TypedExecute.h`, `libreloc/src/TypedExecute.cpp` | Ordered scalar reference execution and qualified kernel adapters |
| Modify | `libreloc/include/reloc/Quant.h`, `libreloc/quant/Quant.cpp`, `libreloc/quant/QuantKernels.h`, `libreloc/quant/QuantAVX2.cpp`, `libreloc/quant/QuantAVX512.cpp` | Existing CPU transform implementations, changing only kernels whose C1 qualification requires it |
| Modify | `libreloc/include/reloc/CudaKernels.h`, `libreloc/cuda/CudaKernels.cu` | CUDA transform qualification and narrowly required missing paths |
| Modify | `libreloc/include/reloc/Transfer.h`, `libreloc/src/Transfer.cpp` (R2) | Typed source/wire/destination descriptors using the same storage/stream ownership contract |
| Modify | `libreloc/include/reloc/Prefold.h`, `libreloc/src/Prefold.cpp` | Backend ownership and typed capability preconditions shared with T4 |
| Create | `libreloc/python/PyDispatch.h`, `libreloc/python/PyDispatch.cpp`, `libreloc/python/reloc_torch/dispatch.py` | Torch-free dispatch binding plus optional tensor/parameter preparation |
| Modify | `libreloc/python/PyReloc.cpp`, `libreloc/python/pyreloc/__init__.py`, `libreloc/python/CMakeLists.txt`, `libreloc/CMakeLists.txt` | Register focused sources and public interfaces |
| Create | `libreloc/test/TypedExecuteTest.cpp`, `libreloc/test/DispatchTest.cpp`, `libreloc/python/tests/torch_frontend/test_dispatch.py` | Numerical cases, capability/policy, real generated typed CPU/CUDA execution |
| Modify | `libreloc/test/CudaKernelsTest.cpp`, `libreloc/test/PrefoldTest.cpp`, `libreloc/test/CMakeLists.txt` | Kernel/prefold contract regression tests |
| Create | `docs/runtime-dispatch.md` | Supported paths, policy, byte accounting, numerical exclusions, evidence |

## Task 1: Execute the initial typed subset with a scalar CPU path

**Interfaces:** Consume C3's immutable typed bound plan, ordered stages, logical descriptors, resolved symbol bindings, and declared parameter bindings. Add a standalone scalar executor for that representation; it consumes explicit validated CPU buffers and produces the declared destination. It must not depend on MLIR, Torch, a GPU, or a cost model.

- [ ] Write analytical C1 cases for f32→f16 rounding boundaries/subnormals/overflow, f16→f32 widening, signed-i8 quantization/dequantization, positive finite scales, allowed zero points, saturation, NaNs/infinities, and signed zero. Use exact expected bytes or C1's stated comparison rule; do not define expected results by calling the kernel being tested.
- [ ] Include per-tensor and per-channel parameters with distinct channel values, a layout that moves the logical channel axis, and equal-sized axes that would conceal a mistaken channel mapping. Test pad→quantize separately from quantize→pad, plus cast/quantize→dequantize sequences that must retain intermediate rounding.
- [ ] Implement ordered scalar stage execution, using checked C3 descriptors and parameter indexes. Keep source/destination scalar types distinct. Supply missing reference operations instead of misusing a layout byte-copy or treating a plain cast as quantization.
- [ ] Reject missing/extra parameters, wrong dtype/shape/channel length, invalid scale/zero-point values, and footprint overflow before data allocation or launch. The first qualified parameter path may require validated CPU-owned values; device-resident values remain explicitly unavailable until Task 4's ordered preparation is implemented.
- [ ] Generate typed plans through C3's real exporter for execution tests; hand-built cases may supplement malformed-input tests. Run new native cases first failing, then passing; commit as `feat(runtime): execute ordered typed plans with scalar semantics`.

## Task 2: Qualify existing kernels against the bound program

**Interfaces:** Add a pure capability query that returns eligible implementation identifiers or a stable exclusion reason for a typed program, direction, bound descriptors, device, parameter metadata, and validated values. No calibration input is needed to answer capability.

- [ ] Create a capability table in `docs/runtime-dispatch.md` for scalar CPU, existing CPU gather/quant and cast paths, CUDA cast/dequant, CUDA dequant/relocate, and existing S8 prefold variants. Record every actual source/destination dtype, stage order, rank, pad, channel mapping, alignment, parameter, and device restriction. Read the implementations before marking a row supported.
- [ ] For each proposed optimized row, compare it against Task 1's analytical cases and the ordered reference program. Check CPU/CUDA differences at ties, saturation, NaNs, subnormals and channel boundaries; test rejected nonzero zero points and padded/rank-mismatched cases if the existing kernel cannot implement them.
- [ ] Add a narrowly scoped missing kernel or retain the correct scalar/original path when a contract mismatch exists. Keep unsupported optimized combinations excluded even when a cost estimate favors them. Do not silently substitute symmetric quantization for a requested affine operation.
- [ ] Define legal stage partitions from C3's ordered program. For example, an explicitly requested f32→i8→f32 program may transfer its i8 intermediate only if both halves implement the same rounding, parameter mapping and stage order. A standalone f32 layout request cannot acquire that intermediate to save bandwidth.
- [ ] Exercise mixed layout+cast and layout+quantize/dequantize using generated artifacts in both directions. Test a non-self-inverse forward D2H program independently; do not apply any inferred inverse value transform.
- [ ] Run native reference and real GPU kernel comparisons; commit as `feat(runtime): qualify typed transform execution paths`.

## Task 3: Dispatch with explicit original and automatic policies

**Interfaces:** In the optional bridge, propose `prepare_typed_transfer(compiled, source, device, *, parameters, policy="auto", calibration=None) -> PreparedTypedTransfer` and `execute_typed_transfer(request) -> DispatchResult`. The prepared request owns the selected immutable program/partition, parameter values/owners, and R2 transport resources. `DispatchResult` contains `tensor` and an immutable scalar-only `report`; T3 can return the tensor and aggregate the report without changing its operator schema.

`policy` accepts `original_cpu` and `auto`. `original_cpu` executes the declared transformations/layout on CPU and performs the necessary cross-device copy: transform then H2D, or D2H of the logical source then forward CPU transformation. This is the forceable runtime baseline. It is distinct from T3's saved-original-callable fallback for a recipe outside compiler/runtime support.

- [ ] Write policy tests where advice prefers an unavailable optimized kernel, calibration is absent, the chosen wire type differs from the final dtype, and the user forces `original_cpu`. Assert exact allowed semantics and the actual invoked implementation, not merely a stored `MethodDecision` enum.
- [ ] Implement the original CPU-transform pipeline using Task 1 and R2. Select it deterministically when auto lacks usable calibration and it is supported; otherwise return an explicit unsupported reason before execution. An unavailable requested baseline must not be reported as successfully forced.
- [ ] For auto, enumerate only Task 2's qualified equivalent alternatives, then consult the existing central cost-model implementation. Translate its advice to an available path or fall back to the qualified original CPU path with a recorded reason. Do not duplicate the model or use benchmark-specific setup/formulas.
- [ ] Derive `source_bytes`, `wire_bytes`, `destination_bytes`, and `parameter_bytes` from the selected checked descriptors and actual partition. Report `payload_bytes_transferred` separately when a path makes extra copies; a representation footprint is not an observed total. Use these exact scalar report field names alongside `implementation`, `policy`, `placement_reason`, and `artifact_version`, with no live tensor references or raw pointers.
- [ ] Verify reports against backend byte/call accounting and kernel invocations. A selected label without execution evidence cannot satisfy this task. Test backend errors propagating without trying another path after data dispatch.
- [ ] Run policy/reference tests and generated CPU/CUDA dispatch tests; commit as `feat(runtime): dispatch qualified plans with an explicit CPU baseline`.

## Task 4: Prepare parameters and qualify prefold ownership for T4

**Interfaces:** Prepared typed requests own validated parameter snapshots or retain owners under a documented immutability/version contract through completion. T4 continues to own `weights.py`, `prefold.py`, the Python prefold bridge, and cache invalidation policy. R3 supplies the runtime capability and lifetime requirements they enforce.

- [ ] Test runtime-supplied CPU parameter values, including replacement and in-place mutation between calls. Revalidate values and recompute capability/preparation as required; an unchanged pointer or tensor reference is insufficient proof that parameters remain valid.
- [ ] Keep device parameter bindings marked unavailable unless they have an explicit preparation path: order after their caller-stream producer, copy into owned host validation storage when needed, wait for validation, and retain the exact validated snapshot/device copy used by execution. This preparation is a separate phase before data-plan dispatch; report its copy/errors accurately. A failed CUDA preparation is an error, not a signal to rerun the original callable. Coordinate any extension of R2's metadata-only preflight boundary with T3 before enabling it.
- [ ] Add tests for invalid scales produced on a CUDA stream, device mismatch, parameter collection, and concurrent replacement of the caller's parameter object. Enable device parameters only after these tests pass on real hardware; otherwise publish the supported CPU-parameter path and stable exclusion.
- [ ] Qualify `APrefold` only when the requested typed program matches an existing S8 output variant and T4 supplies a fresh prepared artifact. Establish ownership that keeps the backend alive until the prefold artifact releases its pinned/device allocations; eliminate reliance on a dangling raw backend pointer in the supported integration.
- [ ] Document T4's required freshness key: program/artifact semantics, bound shape/channel mapping, weight identity/storage/version, parameter identities/values or versions, dtype/device, and any execution-affecting configuration. Require reprepare or fallback after mutation, replacement, `load_state_dict`, or close. Do not introduce a second weight cache in R3.
- [ ] Preserve normal R2 layout preparation for f32 weights; never quantize them merely to use an S8 prefolder. Validate preconditions and backend lifetime in native prefold tests, and hand mutation/slot ownership tests to T4.
- [ ] Run runtime/prefold ownership tests and optional parameter GPU cases; commit as `fix(runtime): own typed parameters and qualify prefold execution`.

## Acceptance and handoff

- [ ] Forced `original_cpu` works for every documented baseline row and is observable; auto chooses only implemented equivalent alternatives.
- [ ] At least one mixed layout+cast and one layout+quantize/dequantize program execute from real compiler artifacts, with numerical and actual-byte evidence.
- [ ] Missing calibration, unsupported kernels, invalid parameters, and execution errors have distinct behavior; none silently changes precision or omits operations.
- [ ] R2 ownership/stream guarantees extend through all typed stages and required parameter lifetimes.
- [ ] C4 receives the dispatch API, capability matrix, exact numerical comparisons and reports; T4 receives prefold capability/ownership requirements without a dependency cycle.

Run in matching configured builds with the real C3 exporter and explicit runtime imports; the GPU command requires actual CUDA execution:

```bash
ctest --test-dir build/sym --output-on-failure
.venv/bin/python -m pytest libreloc/python/tests/torch_frontend/test_dispatch.py -m 'not gpu' -q
.venv/bin/python -m pytest libreloc/python/tests/torch_frontend/test_dispatch.py -m gpu -q
```
